spec.bs

```diff
@@ -6,57 +6,882 @@ Status: CG-DRAFT
+    1. Let |databaseMap| be the result of running [=obtain a shared storage bottle map=] for |environment|.
+    1. Let |count| be |databaseMap|'s [=map/size=].
+    1. Return |count|.
+  </div>
+
+  <div algorithm>
+    To <dfn for="shared storage database">purge expired entries from the database</dfn>, given a [=/browsing context=] |context|, [=enqueue the following steps=] on |context|'s associated [=shared storage database queue=]:
+
+    1. Let |environment| be |context|'s [=active window=]'s [=relevant settings object=].
+    1. Let |databaseMap| be the result of running [=obtain a shared storage bottle map=] for |environment|.
+    1. For each [=entry/key=] |key| in |databaseMap|:
+        1. Let |valueStruct| be the result of running [=map/Get=] on |databaseMap| with |key|.
+        1. Let |isExpired| be the result of running [=shared storage database/determine whether an entry is expired=] with |valueStruct|.
+        1. If |isExpired| is true, [=map/Remove=] |databaseMap|[|key|].
+  </div>
+
+  <div algorithm>
+    To <dfn for="shared storage database">determine whether an entry is expired</dfn>, given a [=entry/value struct=] |valueStruct|:
+
+    1. Let |lastUpdated| be |valueStruct|'s [=value struct/last updated=].
+    1. Let |lifetime| be [=user agent=]'s [=default entry lifetime=].
+    1. Let |expiration| be the sum of |lastUpdated| and |lifetime|.
+    1. let |time| be the current time.
+    1. If |expiration| is less than or equal to |time|, return true.
+    1. Otherwise, return false.
+  </div>
+
+
+
+  ## Monkey Patch for the [=Storage Model=] ## {#monkey-patch}
+  After adding a new [=storage type=] "`shared`" to the [=Storage Model=], this standard will [=storage endpoint/register=] a [=storage endpoint=] of [=storage type|type=] "`shared`" with [=storage identifier=] "`sharedStorage`" and [=storage endpoint/quota=] `5`<sup>`4`</sup> `*` `2`<sup>`16`</sup> bytes (i.e. 39.0625 mebibytes).
+
+  <span class=todo>This [=storage endpoint/quota=] is calculated from the current implementation. Consider bringing the current implementation in line with the spec for [=storage endpoints=] "`localStorage`" and "`sessionStorage`", i.e. `5 * 2`<sup>`20`</sup> bytes. For example, decreasing the per-origin entry limit from 10,000 to 1,280 would achieve this.</span>
+
+  A [=user agent=] holds a [=shared storage shed=] for [=storage endpoints=] of [=storage type|type=] "`shared`".
+
+  A <dfn>shared storage shed</dfn> is a [=map=] of [=/origins=] to to [=storage shelf|storage shelves=]. It is initially empty.
+
+  Note: Unlike [=storage sheds=], whose keys are [=storage keys=], [=shared storage sheds=] use [=/origins=] as keys directly. [=Shared storage=] will be intentionally excluded from [=client-side storage partitioning=].
+
+  A [=user agent=]'s [=shared storage shed=] holds all <dfn>shared storage</dfn> data.
+
+  A <dfn>shared storage shelf</dfn> is a [=storage shelf=] in a [=shared storage shed=].
```

Comment:
Why do we need a new shelf definition?  It seems the existing storage shelf should work?   It will just only have a "default" key in its bucket map because we don't support using shared storage in different storage buckets right now.

Reply:
Changed.